### PR TITLE
21 push notification

### LINKED ISSUE #21: [21pts] Reminders/Push Notifications
**Fill in the blanks**:
As a student, I would like to get reminded when my favorite foods are being served, so that I can prioritize going to DCT over other dining options on certain days.

**Definition of Done - Check as you go!**
- [ ] Tasks have been implemented
- [ ] Documented
- [ ] Passing Tests for each task
- [ ] Pull Request made
- [ ] Code Reviewed

### PR DESCRIPTION
**Pull request by DDVD233**
It can now add a device to a user's device list and notify the user whenever a reminder is required.

Several todos:

- We need a scheduler to send reminders periodically, as sending reminders are not invoked by accessing an API endpoint.
- Frontend needs to request notification permission. We need to have a somewhat friendly interface to ask users for permission.
The todos can be another ticket, separated from this PR to make the PR not too large.

### CODE REVIEW

File: app_backend/app/routers/notification.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    # Update the user
+    db.db["users"].update_one({"_id": user.id}, {"$set": {"devices": devices}})
+
+    resp.status = status.HTTP_200_OK
+    resp.body = {"msg": "Device added successfully"}
+    return resp
+
+
+"""
+    add_device binds a device to a user
```
Comment (RafaelPiloto10) [inline app_backend/app/routers/notification.py]: Update comment

File: app_backend/requirements.txt [hunk 2]
```diff
@@ -31,3 +31,4 @@ typing-extensions==3.10.0.2
 urllib3==1.26.7
 uvicorn==0.15.0
 wrapt==1.12.1
+firebase_admin
```
Comment (RafaelPiloto10) [inline app_backend/requirements.txt:35]: Do we need any auth keys to authenticate?

File: app_backend/app/routers/notification.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    returns Response
+        Response.body: dict
+            Provides any msgs/errs for the request
+        Response.status: int
+            The status code for the request
+"""
+@router.post("/remove_device")
+async def add_device(resp: Response, device_code: str):
+    user = User()  # Todo: Get User from JWT
```
Comment (RafaelPiloto10) [inline app_backend/app/routers/notification.py]: Leaving hanging comment as a reminder

**Review by RafaelPiloto10 — CHANGES_REQUESTED** [review]

**Comment by Foxworth22** [comment]
@oreorz please review this soon

File: app_backend/requirements.txt (hunk 2 above)
Comment (DDVD233) [reply app_backend/requirements.txt:35]: Good point. I will add a Firebase project using the Quack Google account.

Comment (DDVD233) [reply app_backend/requirements.txt:35]: Now I have a private key (a .json file). What's the best way to upload it to the production environment?

File: app_backend/app/routers/notification.py (hunk 3 above)
Comment (DDVD233) [reply app_backend/app/routers/notification.py]: I am not quite sure what this means

**Review by brendacano — CHANGES_REQUESTED** [review]
This looks really good. Please update this branch with the new changes merged into dev because it seems like you will have to make some changes based on that merged ticket.

File: app_backend/app/routers/notification.py (hunk 3 above)
Comment (RafaelPiloto10) [reply app_backend/app/routers/notification.py]: As in this is a reminder to complete this TODO before merging (now that JWT integration is complete) #147

File: app_backend/requirements.txt (hunk 2 above)
Comment (RafaelPiloto10) [reply app_backend/requirements.txt:35]: Just plop it into the project, we can worry about securing our keys later

**Comment by DDVD233** [comment]
Fixed tests; integrated user authentication for both notification and toggle favorite

**Comment by brendacano** [comment]
is this ready for review?

**Review by brendacano — APPROVED** [review]
It looks really good even after having to make changes with the new authentication stuff.

**Review by RafaelPiloto10 — APPROVED** [review]
Everything looks good, just make sure to merge after the parent branch has been merged
#151

**Comment by RafaelPiloto10** [comment]
@DDavid233 it has been merged. Feel free to merge when you get the chance!

**Review by RafaelPiloto10 — APPROVED** [review]

**Comment by RafaelPiloto10** [comment]
@DDavid233 let's merge this ASAP as it has some features we need

**Comment by DDVD233** [comment]
@RafaelPiloto10 Merged

**Pull request merged**